### ISSUE TITLE
jgit cloning not converting line ends on windows?

### ISSUE THREAD

**Issue by jtnord** [jenkins-infra/helpdesk#3865]
### Service(s)

ci.jenkins.io

### Summary

https://github.com/jenkinsci/google-oauth-plugin/pull/209/ fails in [CI on windows ](https://ci.jenkins.io/job/Plugins/job/google-oauth-plugin/job/PR-209/1/execution/node/154/log/) as it appears all the files have UNIX line endings.

https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fgoogle-oauth-plugin/detail/PR-209/1/pipeline/79/#step-98-log-14 appears to show that the clone/fetch of the project is done by jgit and not git.
this could be a case of 
a) misreporting by sportless (but it affects 50 files which is every single java file in src/main) 
b) spotless is not pickup up where the global/system config should be and is assuming they should be \n and are \r\n (they have been converted)
c) jgit is not pickup up where the global/system config should be and not converting the files
d) the option is not configured - so there is no conversion and spotless falls back to expecting native line end.
e) something is very messed up in the project (but it works locally on windows).
f) jgit plugin ignores settings from the agent and need to be specified in code.

without some access to the node it is not possible to diagnose for a plugin developer what is wrong here, but it seems to not be completely widespread.

**Comment by jtnord** [comment]
was worked around by switching to container agents - so this could be something specific to the VMs (in Azure?)

**Comment by dduportal** [comment]
> was worked around by switching to container agents - so this could be something specific to the VMs (in Azure?)

Thanks for the pointer (sorry for the delay, today was really busy for the whole team).

It make sense as we have not (yet) unified Windows VM and container images:

- Container images are "classic" Dockerfiles defined in https://github.com/jenkins-infra/docker-inbound-agents/tree/main/maven
- While VM are defined in https://github.com/jenkins-infra/packer-images

**Comment by jonesbusy** [comment]
Similar issue here : https://ci.jenkins.io/job/Plugins/job/locale-plugin/job/main/83/console after I formatted the plugin and updated Jenkinsfile

Was a mistake to build on VM. Fixed by using `useContainerAgent: true`.

So behavior is definitely not the same between Windows VM and container agent

**Comment by jonesbusy** [comment]
Other occurence on this repo https://github.com/jenkins-infra/plugin-modernizer-tool/pull/464/checks?check_run_id=34452658438

**Comment by jtnord** [comment]
> @jtnord I do not know which setting to look for: I guess since it is `jgit` it could be a general Windows setting, a git (installlation) setting or even a Java setting?

Sorry I missed the ping.

the setting is probably `core.autocrlf` in the global/system git config.  
as this is probably not set, I would guess that jgit is using a default of false and so not doing conversion, spotless is then looking up the defaults, not seeing any conversion or specific line settings and going OK the default is platform native..

https://github.com/diffplug/spotless/issues/1559

**Comment by MarkEWaite** [comment]
> for code that does not care about windows users preference for linux or windows either way, a workaround in the interim is to add a `.gitattributes` and set all text files to have linux line ends (rather than auto) (`* text eol=lf`). (which spotless honours)

Thanks!  I've used that technique with several of my plugins, but had to use a more specific pattern match so that it would not treat binary files as text.  The [git plugin .gitattributes file](https://github.com/jenkinsci/git-plugin/blob/master/.gitattributes) and the [git client plugin .gitattributes file](https://github.com/jenkinsci/git-client-plugin/blob/master/.gitattributes) are two examples.

However, I'm not sure if we want to make that a general pattern throughout Jenkins or if that is just reinforcing my bias towards Unix line endings.

**Comment by MarkEWaite** [comment]
The `core.autocrlf` setting seems to have resolved the issue.  I propose that we wait a few days to confirm it, then close this issue.

**Comment by jtnord** [comment]
>  I propose that we wait a few days to confirm it, then close this issue.

I would suggest that the git settings should be in the VM if Jenkins needs them, not in the pipeline library.

Otherwise other jobs (non plugins) that use windows VMs will also be affected.

**Comment by dduportal** [comment]
> > I propose that we wait a few days to confirm it, then close this issue.
> 
> I would suggest that the git settings should be in the VM if Jenkins needs them, not in the pipeline library.
> 
> Otherwise other jobs (non plugins) that use windows VMs will also be affected.

Agreed with James! Fixing at the VM template is easy and would solve the problem once for all without requiring per-project change. Incoming PRs

**Comment by dduportal** [comment]
PR https://github.com/jenkins-infra/packer-images/pull/1649 opened to change the default value core Git's `core.autocrlf` in the Windows VMs from `false` to `true`

**Comment by dduportal** [comment]
Deployed with the latest (https://github.com/jenkins-infra/packer-images/releases/tag/2.21.0) version of the VM template ! Thanks again @jtnord and @MarkEWaite
